File: language/real/prolog/challenge_v14.pl
:- use_module(library(readutil)).
:- use_module(library(ordsets)).

:- dynamic packet_id/1, claim_scope/1, candidate_relations/1.
:- dynamic selector/2, selection_timing/1, selection_rule/3.
:- dynamic route/1, probe/5, covers/2, score_source/2, relevance_source/2, expansion/2.
:- dynamic expected_authority/1, expected_narrowed/1, expected_coverage/1.
:- dynamic expected_common/1, expected_capture/1, expected_expansion_reopen/1.
:- dynamic seen_header/0, seen_end/0.

relations(['DISJOINT','EXACT','MERGE','OVERLAP','REFINE','UNMAPPED']).

reset_db :-
  retractall(packet_id(_)),retractall(claim_scope(_)),retractall(candidate_relations(_)),
  retractall(selector(_,_)),retractall(selection_timing(_)),retractall(selection_rule(_,_,_)),
  retractall(route(_)),retractall(probe(_,_,_,_,_)),retractall(covers(_,_)),
  retractall(score_source(_,_)),retractall(relevance_source(_,_)),retractall(expansion(_,_)),
  retractall(expected_authority(_)),retractall(expected_narrowed(_)),retractall(expected_coverage(_)),
  retractall(expected_common(_)),retractall(expected_capture(_)),retractall(expected_expansion_reopen(_)),
  retractall(seen_header),retractall(seen_end).

atomize(S,A):-atom_string(A,S).
plus_relations(S,R):-split_string(S,"+","",Ps),Ps\=[],maplist(atomize,Ps,A0),maplist(upcase_atom,A0,A1),sort(A1,R),relations(All),ord_subset(R,All).

parse_line(Line):-
  normalize_space(string(N),Line),
  (N=""->true
  ;sub_string(N,0,1,_,"#")->true
  ;seen_end->true
  ;split_string(N," \t"," \t",T),
    (T=["REALCHALLENGE","0.14"]->parse_tokens(T)
    ;seen_header->parse_tokens(T)
    ;format(user_error,'PROLOG_HEADER_REQUIRED ~s~n',[Line]),fail)).

parse_tokens(["REALCHALLENGE","0.14"]):- \+seen_header,assertz(seen_header).
parse_tokens(["id",S]):-atomize(S,A),assertz(packet_id(A)).
parse_tokens(["claim_scope",S]):-atomize(S,A),assertz(claim_scope(A)).
parse_tokens(["candidate_relations",S]):-plus_relations(S,R),assertz(candidate_relations(R)).
parse_tokens(["selector",I,A]):-atomize(I,AI),atomize(A,AA),assertz(selector(AI,AA)).
parse_tokens(["selection_timing",S]):-atomize(S,A0),upcase_atom(A0,A),member(A,['PRESEALED','POST_OUTCOME']),assertz(selection_timing(A)).
parse_tokens(["selection_rule",I,A,M]):-atomize(I,AI),atomize(A,AA),atomize(M,M0),upcase_atom(M0,AM),member(AM,['INDEPENDENT','INCUMBENT_DEFINED','UNRESOLVED']),assertz(selection_rule(AI,AA,AM)).
parse_tokens(["route",I]):-atomize(I,A),\+route(A),assertz(route(A)).
parse_tokens(["probe",I,A,S,M,R]):-atomize(I,AI),atomize(A,AA),atomize(S,S0),upcase_atom(S0,AS),member(AS,['SELECTED','UNSELECTED']),atomize(M,M0),upcase_atom(M0,AM),member(AM,['WORLD_FACING','INTERNAL']),plus_relations(R,Rs),\+probe(AI,_,_,_,_),assertz(probe(AI,AA,AS,AM,Rs)).
parse_tokens(["covers",P,R]):-atomize(P,AP),atomize(R,AR),assertz(covers(AP,AR)).
parse_tokens(["score_source",P,S]):-atomize(P,AP),atomize(S,S0),upcase_atom(S0,AS),member(AS,['INDEPENDENT','CANDIDATE_DERIVED','STANDARD_DERIVED']),assertz(score_source(AP,AS)).
parse_tokens(["relevance_source",P,S]):-atomize(P,AP),atomize(S,S0),upcase_atom(S0,AS),member(AS,['INDEPENDENT','CANDIDATE_DERIVED','STANDARD_DERIVED']),assertz(relevance_source(AP,AS)).
parse_tokens(["expansion",I,R]):-atomize(I,AI),plus_relations(R,Rs),assertz(expansion(AI,Rs)).
parse_tokens(["authorize_selection_authority",S]):-atomize(S,A0),upcase_atom(A0,A),assertz(expected_authority(A)).
parse_tokens(["authorize_narrowed",S]):-atomize(S,A0),upcase_atom(A0,A),assertz(expected_narrowed(A)).
parse_tokens(["authorize_declared_route_coverage_complete",S]):-atomize(S,A0),upcase_atom(A0,A),assertz(expected_coverage(A)).
parse_tokens(["authorize_common_ancestry",S]):-atomize(S,A0),upcase_atom(A0,A),assertz(expected_common(A)).
parse_tokens(["authorize_discriminator_capture",S]):-atomize(S,A0),upcase_atom(A0,A),assertz(expected_capture(A)).
parse_tokens(["authorize_expansion_reopen",S]):-atomize(S,A0),upcase_atom(A0,A),assertz(expected_expansion_reopen(A)).
parse_tokens(["END"]):-assertz(seen_end).

load_packet(File):-
  reset_db,read_file_to_string(File,S,[]),split_string(S,"\n","\r",Ls),maplist(parse_line,Ls),
  seen_header,seen_end,packet_id(_),claim_scope(_),candidate_relations(_),selector(_,_),
  selection_timing(_),selection_rule(_,_,_),route(_),probe(_,_,_,_,_),
  \+(covers(P,_),\+probe(P,_,_,_,_)),
  \+(covers(_,R),\+route(R)),
  \+(probe(P,_,_,_,_),\+score_source(P,_)),
  \+(probe(P,_,_,_,_),\+relevance_source(P,_)).

selected_world(P,A,R):-probe(P,A,'SELECTED','WORLD_FACING',R).
selected_any(P):-probe(P,_,'SELECTED',_,_).

intersect_all([H|T],R):-foldl(ord_intersection,T,H,R).
after_set(A):-
  candidate_relations(B),
  findall(S,selected_world(_,_,S),Ss),
  (Ss=[]->A=B;intersect_all([B|Ss],A)).

covered_selected_route(R):-covers(P,R),selected_world(P,_,_).

common_ancestry:-
  findall(P,selected_world(P,_,_),Ps0),sort(Ps0,Ps),length(Ps,N),
  findall(A,selected_world(_,A,_),As0),sort(As0,As),length(As,M),
  N>1,N>M.

incumbent_relevance:-selection_rule(_,_, 'INCUMBENT_DEFINED'),!.
incumbent_relevance:-selected_world(P,_,_),relevance_source(P,'CANDIDATE_DERIVED'),!.

candidate_score:-selected_world(P,_,_),score_source(P,'CANDIDATE_DERIVED'),!.
discriminator_capture:-incumbent_relevance,!.
discriminator_capture:-candidate_score,!.

omitted_counterprobe:-
  after_set(A),A\=[],
  probe(P,_, 'UNSELECTED','WORLD_FACING',S),
  candidate_relations(B),ord_intersection(B,S,X),X\=[],
  ord_intersection(A,S,[]),!.

expansion_result(S,X):-after_set(A),expansion(_,S),ord_intersection(A,S,X).
expansion_instability:-after_set(A),expansion_result(_,X),X\=A,!.
expansion_reopen:-after_set(A),A\=[],expansion_result(_,[]),!.
expansion_narrow:-after_set(A),expansion_result(_,X),X\=[],length(X,NX),length(A,NA),NX<NA,!.

yes(true,'YES'):-!.
yes(_,'NO').

authority_state(Post,Capture,Omitted,WorldN,Coverage,Common,ExpansionReopen,After,Narrow,Authority):-
  (Post=true->Authority='HOLD_POST_OUTCOME_SELECTION'
  ;Capture=true->Authority='HOLD_DISCRIMINATOR_CAPTURE'
  ;Omitted=true->Authority='REOPEN_OMITTED_COUNTERPROBE'
  ;WorldN=:=0->Authority='HOLD_NO_WORLD_FACING'
  ;Coverage=false->Authority='HOLD_ROUTE_COVERAGE'
  ;Common=true->Authority='HOLD_COMMON_ANCESTRY'
  ;ExpansionReopen=true->Authority='REOPEN_EXPANSION_CONFLICT'
  ;After=[]->Authority='REOPEN_SELECTED_CONFLICT'
  ;Narrow=true->Authority='AUTHORIZED_PROVISIONAL_NARROWING'
  ;Authority='ADMISSIBLE_NO_NARROWING').

check_expected(A,N,C,Com,Cap,ER):-
  (expected_authority(X)->X=A;true),
  (expected_narrowed(X)->X=N;true),
  (expected_coverage(X)->X=C;true),
  (expected_common(X)->X=Com;true),
  (expected_capture(X)->X=Cap;true),
  (expected_expansion_reopen(X)->X=ER;true).

run(File):-
  load_packet(File),
  findall(P,probe(P,_,_,_,_),Reg0),sort(Reg0,Reg),length(Reg,RegN),
  findall(P,selected_any(P),Sel0),sort(Sel0,Sel),length(Sel,SelN),
  findall(P,selected_world(P,_,_),W0),sort(W0,W),length(W,WorldN),
  findall(A,selected_world(_,A,_),As0),sort(As0,As),length(As,AncN),
  findall(R,route(R),Rs0),sort(Rs0,Rs),length(Rs,RouteN),
  findall(R,covered_selected_route(R),CR0),sort(CR0,CR),length(CR,CoverN),
  UncoverN is RouteN-CoverN,(UncoverN=:=0->CoverageB=true;CoverageB=false),yes(CoverageB,Coverage),
  (WorldN>1,WorldN>CoverN->MultB=true;MultB=false),yes(MultB,Mult),
  (common_ancestry->CommonB=true;CommonB=false),yes(CommonB,Common),
  (selection_timing('POST_OUTCOME')->PostB=true;PostB=false),yes(PostB,Post),
  (incumbent_relevance->IncB=true;IncB=false),yes(IncB,Inc),
  (discriminator_capture->CapB=true;CapB=false),yes(CapB,Capture),
  (omitted_counterprobe->OmitB=true;OmitB=false),yes(OmitB,Omitted),
  candidate_relations(Before),after_set(After),
  length(Before,NB),length(After,NA),(After\=[],NA<NB->NarrowB=true;NarrowB=false),yes(NarrowB,Narrow),
  findall(I,expansion(I,_),Es0),sort(Es0,Es),length(Es,ExpN),
  (expansion_instability->EIB=true;EIB=false),yes(EIB,EI),
  (expansion_reopen->ERB=true;ERB=false),yes(ERB,ER),
  (expansion_narrow->ENB=true;ENB=false),
  (ExpN=:=0->EState='NONE';ERB=true->EState='REOPENED';ENB=true->EState='NARROWED';EIB=true->EState='CHANGED';EState='STABLE'),
  authority_state(PostB,CapB,OmitB,WorldN,CoverageB,CommonB,ERB,After,NarrowB,Authority),
  atomic_list_concat(Before,'+',BeforeA),
  (After=[]->AfterA='NONE';atomic_list_concat(After,'+',AfterA)),
  format('challenge.registered_probe_count=~w~n',[RegN]),
  format('challenge.selected_probe_count=~w~n',[SelN]),
  format('challenge.selected_world_probe_count=~w~n',[WorldN]),
  format('challenge.selected_ancestry_count=~w~n',[AncN]),
  format('challenge.route_count=~w~n',[RouteN]),
  format('challenge.covered_route_count=~w~n',[CoverN]),
  format('challenge.uncovered_route_count=~w~n',[UncoverN]),
  format('challenge.declared_route_coverage_complete=~w~n',[Coverage]),
  format('challenge.probe_route_multiplicity=~w~n',[Mult]),
  format('challenge.common_ancestry=~w~n',[Common]),
  format('challenge.post_outcome_selection=~w~n',[Post]),
  format('challenge.incumbent_relevance=~w~n',[Inc]),
  format('challenge.discriminator_capture=~w~n',[Capture]),
  format('challenge.omitted_counterprobe=~w~n',[Omitted]),
  format('challenge.correspondence_before=~w~n',[BeforeA]),
  format('challenge.correspondence_after=~w~n',[AfterA]),
  format('challenge.narrowed=~w~n',[Narrow]),
  format('challenge.expansion_count=~w~n',[ExpN]),
  format('challenge.expansion_instability=~w~n',[EI]),
  format('challenge.expansion_reopen_required=~w~n',[ER]),
  format('challenge.expansion_state=~w~n',[EState]),
  writeln('challenge.current_family_complete=NO'),
  writeln('challenge.future_challenge_space_closed=NO'),
  writeln('challenge.probe_count_truth_oracle=NO'),
  writeln('challenge.externality_truth_oracle=NO'),
  writeln('challenge.cost_truth_oracle=NO'),
  writeln('challenge.selection_rule_truth_oracle=NO'),
  format('challenge.selection_authority=~w~n',[Authority]),
  writeln('challenge.guidance_mode=CONSTITUTED_REOPENABLE_SEPARATOR_AUTHORITY'),
  check_expected(Authority,Narrow,Coverage,Common,Capture,ER).

File: language/real/src/challenge_v14.rs
use std::{
    collections::{BTreeMap, BTreeSet},
    env, fs,
    path::Path,
};

const RELATIONS: [&str; 6] = ["DISJOINT","EXACT","MERGE","OVERLAP","REFINE","UNMAPPED"];

#[derive(Clone, Debug)]
struct Probe {
    id: String,
    ancestry: String,
    selected: bool,
    mode: String,
    supports: BTreeSet<String>,
}

#[derive(Default, Debug)]
struct Packet {
    id: String,
    claim_scope: String,
    candidate_relations: BTreeSet<String>,
    selector_id: String,
    selector_ancestry: String,
    selection_timing: String,
    rule_id: String,
    rule_ancestry: String,
    rule_mode: String,
    routes: BTreeSet<String>,
    probes: BTreeMap<String, Probe>,
    covers: Vec<(String,String)>,
    score_source: BTreeMap<String,String>,
    relevance_source: BTreeMap<String,String>,
    expansions: Vec<(String,BTreeSet<String>)>,
    expected_authority: Option<String>,
    expected_narrowed: Option<String>,
    expected_coverage: Option<String>,
    expected_common: Option<String>,
    expected_capture: Option<String>,
    expected_expansion_reopen: Option<String>,
}

fn toks(s:&str)->Vec<String>{
    s.split_whitespace().map(|x|x.trim_matches('"').to_string()).collect()
}

fn split_plus(s:&str)->Result<BTreeSet<String>,String>{
    let xs:BTreeSet<String>=s.split('+').filter(|x|!x.is_empty()).map(|x|x.to_uppercase()).collect();
    if xs.is_empty(){return Err("empty relation set".into());}
    for x in &xs{
        if !RELATIONS.contains(&x.as_str()){return Err(format!("invalid relation: {x}"));}
    }
    Ok(xs)
}

fn yes(x:bool)->&'static str{if x{"YES"}else{"NO"}}

fn join(xs:&BTreeSet<String>)->String{
    if xs.is_empty(){"NONE".into()}else{xs.iter().cloned().collect::<Vec<_>>().join("+")}
}

fn parse(path:&Path)->Result<Packet,String>{
    let text=fs::read_to_string(path).map_err(|e|e.to_string())?;
    let mut p=Packet::default();
    let mut started=false;
    let mut ended=false;

    for (i,raw) in text.lines().enumerate(){
        let line=raw.trim();
        if line.is_empty()||line.starts_with('#'){continue;}
        let t=toks(line);
        if t.is_empty(){continue;}
        if t[0]=="REALCHALLENGE"{
            if t.len()!=2||t[1]!="0.14"{return Err("expected REALCHALLENGE 0.14".into());}
            if started{return Err("duplicate header".into());}
            started=true;
            continue;
        }
        if !started{return Err("packet must begin REALCHALLENGE 0.14".into());}
        if t[0]=="END"{
            if t.len()!=1{return Err("END takes no arguments".into());}
            ended=true;
            break;
        }
        match t[0].as_str(){
            "id" if t.len()==2=>p.id=t[1].clone(),
            "claim_scope" if t.len()==2=>p.claim_scope=t[1].clone(),
            "candidate_relations" if t.len()==2=>p.candidate_relations=split_plus(&t[1])?,
            "selector" if t.len()==3=>{
                p.selector_id=t[1].clone(); p.selector_ancestry=t[2].clone();
            }
            "selection_timing" if t.len()==2=>{
                let x=t[1].to_uppercase();
                if !["PRESEALED","POST_OUTCOME"].contains(&x.as_str()){return Err("invalid selection timing".into());}
                p.selection_timing=x;
            }
            "selection_rule" if t.len()==4=>{
                let x=t[3].to_uppercase();
                if !["INDEPENDENT","INCUMBENT_DEFINED","UNRESOLVED"].contains(&x.as_str()){return Err("invalid selection rule mode".into());}
                p.rule_id=t[1].clone(); p.rule_ancestry=t[2].clone(); p.rule_mode=x;
            }
            "route" if t.len()==2=>{
                if !p.routes.insert(t[1].clone()){return Err(format!("duplicate route: {}",t[1]));}
            }
            "probe" if t.len()==6=>{
                let selected=match t[3].to_uppercase().as_str(){
                    "SELECTED"=>true,"UNSELECTED"=>false,_=>return Err("invalid probe selection state".into())
                };
                let mode=t[4].to_uppercase();
                if !["WORLD_FACING","INTERNAL"].contains(&mode.as_str()){return Err("invalid probe mode".into());}
                if p.probes.contains_key(&t[1]){return Err(format!("duplicate probe: {}",t[1]));}
                p.probes.insert(t[1].clone(),Probe{
                    id:t[1].clone(), ancestry:t[2].clone(), selected, mode, supports:split_plus(&t[5])?
                });
            }
            "covers" if t.len()==3=>p.covers.push((t[1].clone(),t[2].clone())),
            "score_source" if t.len()==3=>{
                let x=t[2].to_uppercase();
                if !["INDEPENDENT","CANDIDATE_DERIVED","STANDARD_DERIVED"].contains(&x.as_str()){return Err("invalid score source".into());}
                p.score_source.insert(t[1].clone(),x);
            }
            "relevance_source" if t.len()==3=>{
                let x=t[2].to_uppercase();
                if !["INDEPENDENT","CANDIDATE_DERIVED","STANDARD_DERIVED"].contains(&x.as_str()){return Err("invalid relevance source".into());}
                p.relevance_source.insert(t[1].clone(),x);
            }
            "expansion" if t.len()==3=>p.expansions.push((t[1].clone(),split_plus(&t[2])?)),
            "authorize_selection_authority" if t.len()==2=>p.expected_authority=Some(t[1].to_uppercase()),
            "authorize_narrowed" if t.len()==2=>p.expected_narrowed=Some(t[1].to_uppercase()),
            "authorize_declared_route_coverage_complete" if t.len()==2=>p.expected_coverage=Some(t[1].to_uppercase()),
            "authorize_common_ancestry" if t.len()==2=>p.expected_common=Some(t[1].to_uppercase()),
            "authorize_discriminator_capture" if t.len()==2=>p.expected_capture=Some(t[1].to_uppercase()),
            "authorize_expansion_reopen" if t.len()==2=>p.expected_expansion_reopen=Some(t[1].to_uppercase()),
            _=>return Err(format!("{}:{} malformed command: {line}",path.display(),i+1)),
        }
    }

    if !ended{return Err("missing END".into());}
    if p.id.is_empty()||p.claim_scope.is_empty()||p.candidate_relations.is_empty()
        ||p.selector_id.is_empty()||p.selector_ancestry.is_empty()||p.selection_timing.is_empty()
        ||p.rule_id.is_empty()||p.rule_ancestry.is_empty()||p.rule_mode.is_empty()
        ||p.routes.is_empty()||p.probes.is_empty(){
        return Err("missing required field".into());
    }
    for (probe,route) in &p.covers{
        if !p.probes.contains_key(probe){return Err(format!("coverage references unknown probe: {probe}"));}
        if !p.routes.contains(route){return Err(format!("coverage references unknown route: {route}"));}
    }
    for id in p.probes.keys(){
        if !p.score_source.contains_key(id){return Err(format!("missing score_source for probe: {id}"));}
        if !p.relevance_source.contains_key(id){return Err(format!("missing relevance_source for probe: {id}"));}
    }
    Ok(p)
}

fn intersect(a:&BTreeSet<String>,b:&BTreeSet<String>)->BTreeSet<String>{
    a.intersection(b).cloned().collect()
}

fn analyze(p:&Packet)->BTreeMap<String,String>{
    let selected_world:Vec<&Probe>=p.probes.values().filter(|q|q.selected&&q.mode=="WORLD_FACING").collect();
    let selected_count=p.probes.values().filter(|q|q.selected).count();
    let selected_world_count=selected_world.len();
    let selected_ancestries:BTreeSet<String>=selected_world.iter().map(|q|q.ancestry.clone()).collect();
    let selected_ancestry_count=selected_ancestries.len();
    let common_ancestry=selected_world_count>selected_ancestry_count && selected_world_count>1;

    let mut covered_routes=BTreeSet::new();
    for (probe,route) in &p.covers{
        if let Some(q)=p.probes.get(probe){
            if q.selected && q.mode=="WORLD_FACING"{covered_routes.insert(route.clone());}
        }
    }
    let route_count=p.routes.len();
    let covered_route_count=covered_routes.len();
    let uncovered_route_count=route_count.saturating_sub(covered_route_count);
    let coverage_complete=uncovered_route_count==0;
    let probe_route_multiplicity=selected_world_count>covered_route_count && selected_world_count>1;

    let post_outcome=p.selection_timing=="POST_OUTCOME";
    let incumbent_relevance=p.rule_mode=="INCUMBENT_DEFINED" || selected_world.iter().any(|q|p.relevance_source.get(&q.id).map(|x|x=="CANDIDATE_DERIVED").unwrap_or(false));
    let candidate_score=selected_world.iter().any(|q|p.score_source.get(&q.id).map(|x|x=="CANDIDATE_DERIVED").unwrap_or(false));
    let discriminator_capture=incumbent_relevance||candidate_score;

    let before=p.candidate_relations.clone();
    let mut after=before.clone();
    for q in &selected_world{after=intersect(&after,&q.supports);}
    let narrowed=!after.is_empty()&&after.len()<before.len();

    let omitted_counterprobe=p.probes.values().any(|q|{
        if q.selected||q.mode!="WORLD_FACING"||after.is_empty(){return false;}
        intersect(&after,&q.supports).is_empty() && !intersect(&before,&q.supports).is_empty()
    });

    let mut expansion_instability=false;
    let mut expansion_reopen=false;
    let mut expansion_narrow=false;
    for (_,support) in &p.expansions{
        let x=intersect(&after,support);
        if x!=after{expansion_instability=true;}
        if !after.is_empty()&&x.is_empty(){expansion_reopen=true;}
        if !x.is_empty()&&x.len()<after.len(){expansion_narrow=true;}
    }
    let expansion_state=if p.expansions.is_empty(){"NONE"}
        else if expansion_reopen{"REOPENED"}
        else if expansion_narrow{"NARROWED"}
        else if expansion_instability{"CHANGED"}
        else{"STABLE"};

    let authority=if post_outcome{"HOLD_POST_OUTCOME_SELECTION"}
        else if discriminator_capture{"HOLD_DISCRIMINATOR_CAPTURE"}
        else if omitted_counterprobe{"REOPEN_OMITTED_COUNTERPROBE"}
        else if selected_world_count==0{"HOLD_NO_WORLD_FACING"}
        else if !coverage_complete{"HOLD_ROUTE_COVERAGE"}
        else if common_ancestry{"HOLD_COMMON_ANCESTRY"}
        else if expansion_reopen{"REOPEN_EXPANSION_CONFLICT"}
        else if after.is_empty(){"REOPEN_SELECTED_CONFLICT"}
        else if narrowed{"AUTHORIZED_PROVISIONAL_NARROWING"}
        else{"ADMISSIBLE_NO_NARROWING"};

    let mut a=BTreeMap::new();
    a.insert("challenge.registered_probe_count".into(),p.probes.len().to_string());
    a.insert("challenge.selected_probe_count".into(),selected_count.to_string());
    a.insert("challenge.selected_world_probe_count".into(),selected_world_count.to_string());
    a.insert("challenge.selected_ancestry_count".into(),selected_ancestry_count.to_string());
    a.insert("challenge.route_count".into(),route_count.to_string());
    a.insert("challenge.covered_route_count".into(),covered_route_count.to_string());
    a.insert("challenge.uncovered_route_count".into(),uncovered_route_count.to_string());
    a.insert("challenge.declared_route_coverage_complete".into(),yes(coverage_complete).into());
    a.insert("challenge.probe_route_multiplicity".into(),yes(probe_route_multiplicity).into());
    a.insert("challenge.common_ancestry".into(),yes(common_ancestry).into());
    a.insert("challenge.post_outcome_selection".into(),yes(post_outcome).into());
    a.insert("challenge.incumbent_relevance".into(),yes(incumbent_relevance).into());
    a.insert("challenge.discriminator_capture".into(),yes(discriminator_capture).into());
    a.insert("challenge.omitted_counterprobe".into(),yes(omitted_counterprobe).into());
    a.insert("challenge.correspondence_before".into(),join(&before));
    a.insert("challenge.correspondence_after".into(),join(&after));
    a.insert("challenge.narrowed".into(),yes(narrowed).into());
    a.insert("challenge.expansion_count".into(),p.expansions.len().to_string());
    a.insert("challenge.expansion_instability".into(),yes(expansion_instability).into());
    a.insert("challenge.expansion_reopen_required".into(),yes(expansion_reopen).into());
    a.insert("challenge.expansion_state".into(),expansion_state.into());
    a.insert("challenge.current_family_complete".into(),"NO".into());
    a.insert("challenge.future_challenge_space_closed".into(),"NO".into());
    a.insert("challenge.probe_count_truth_oracle".into(),"NO".into());
    a.insert("challenge.externality_truth_oracle".into(),"NO".into());
    a.insert("challenge.cost_truth_oracle".into(),"NO".into());
    a.insert("challenge.selection_rule_truth_oracle".into(),"NO".into());
    a.insert("challenge.selection_authority".into(),authority.into());
    a.insert("challenge.guidance_mode".into(),"CONSTITUTED_REOPENABLE_SEPARATOR_AUTHORITY".into());
    a
}

fn validate(p:&Packet,a:&BTreeMap<String,String>)->Result<(),String>{
    let checks=[
        (&p.expected_authority,"challenge.selection_authority"),
        (&p.expected_narrowed,"challenge.narrowed"),
        (&p.expected_coverage,"challenge.declared_route_coverage_complete"),
        (&p.expected_common,"challenge.common_ancestry"),
        (&p.expected_capture,"challenge.discriminator_capture"),
        (&p.expected_expansion_reopen,"challenge.expansion_reopen_required"),
    ];
    for (e,k) in checks{
        if let Some(x)=e{
            if a[k]!=*x{return Err(format!("{k} mismatch expected {x} got {}",a[k]));}
        }
    }
    Ok(())
}

fn emit(p:&Packet,a:&BTreeMap<String,String>)->String{
    let mut s=format!("REAL-CHALLENGE=0.14\nid={}\nclaim_scope={}\n",p.id,p.claim_scope);
    for k in [
        "challenge.registered_probe_count",
        "challenge.selected_probe_count",
        "challenge.selected_world_probe_count",
        "challenge.selected_ancestry_count",
        "challenge.route_count",
        "challenge.covered_route_count",
        "challenge.uncovered_route_count",
        "challenge.declared_route_coverage_complete",
        "challenge.probe_route_multiplicity",
        "challenge.common_ancestry",
        "challenge.post_outcome_selection",
        "challenge.incumbent_relevance",
        "challenge.discriminator_capture",
        "challenge.omitted_counterprobe",
        "challenge.correspondence_before",
        "challenge.correspondence_after",
        "challenge.narrowed",
        "challenge.expansion_count",
        "challenge.expansion_instability",
        "challenge.expansion_reopen_required",
        "challenge.expansion_state",
        "challenge.current_family_complete",
        "challenge.future_challenge_space_closed",
        "challenge.probe_count_truth_oracle",
        "challenge.externality_truth_oracle",
        "challenge.cost_truth_oracle",
        "challenge.selection_rule_truth_oracle",
        "challenge.selection_authority",
        "challenge.guidance_mode",
    ]{s.push_str(&format!("{k}={}\n",a[k]));}
    s
}

fn main(){
    let args:Vec<String>=env::args().skip(1).collect();
    if args.is_empty(){eprintln!("usage: real-v14-challenge <packet.real> [...]");std::process::exit(2);}
    for f in args{
        match parse(Path::new(&f)){
            Ok(p)=>{
                let a=analyze(&p);
                if let Err(e)=validate(&p,&a){eprintln!("{e}");std::process::exit(1);}
                print!("{}",emit(&p,&a));
            }
            Err(e)=>{eprintln!("{e}");std::process::exit(1);}
        }
    }
}
